### DISCUSSION TITLE
Dependency injection without resorting to global state

### DISCUSSION THREAD

**Discussion by codethief** [fastapi/fastapi#8968] [Questions] [answered]
Hi there! I'm starting the FastAPI web server via Uvicorn after completing various initialization steps (reading command line arguments, reading a configuration file, setting up shared memory etc.). Now I would like to access said dependencies (command line arguments, config, shared memory, …) in my endpoints or, more specifically, pass them as argument to my endpoints. Unfortunately, the current way of injecting dependencies into endpoints via `Depends` basically amounts to using global state, as the dependency has to be declared in the endpoint's function signature as a *Callable* and so has to be available already when Python runs the semantic analysis on the code.

At the same time, `app.include_router(my_router, additional_dependencies=[ ])` does not solve this issue, either, as AFAIK there is no way to access these dependencies inside my endpoints.

What's the best practice to solve this issue? Ideally I would like to do something like

```python
from dataclasses import dataclass

from fastapi import FastAPI, APIRouter

@dataclass
class MyConfig:
    foo: str = ""

router = APIRouter()

@router.post("/")
def my_endpoint(my_config: MyConfig = DependsOnType(MyConfig)):
    ...

def create_app(config: MyConfig) -> FastAPI:
    app = FastAPI()
    app.register_dependency(type=MyConfig, value=config)
    app.include_router(router, prefix="/myrouter")
    return app
```

Then I could do `app = create_app(MyConfig(foo="bar"))` to create different versions of the app, depending on the configuration and other dependencies.

EDIT: While there is a way to achieve this (see Kludex's comment below), it has various deficiencies (see the discussion below) and I would like to propose the above functionality involving `DependsOnType` and `register_dependency()` as a new feature.

**Comment by Kludex** [comment]
The last two sections can help you: https://www.starlette.io/applications/#storing-state-on-the-app-instance

Store in `app.state`, and get from `request.app`.

**Comment by codethief** [comment]
Ohh, thank you so much!

I think my suggestion from above involving `DependsOnType()` would still be desirable, though, because monkey-patching `app.state`…

- breaks type hints as the type of `app.state` is unclear to endpoints.
- it makes all dependencies available to *all* endpoints which might be dangerous.
- it hides an endpoint's dependencies because the module containing the endpoint does not even need to import the required type.

I have therefore converted my question into a feature request.

**Comment by Mause** [comment]
In essence you want a singleton, which isn't currently supported by the dependency injection system in fastapi.

I wrote a quick implementation for my own system though, you're welcome to use it: https://github.com/Mause/media/blob/master/rarbg_local/singleton.py#L33

**Comment by codethief** [comment]
I don't want a singleton, at least not in the strict sense. I just want to pass an instance of an arbitrary class (whether it's a singleton class or not) to all the endpoints in my FastAPI app.

As for your implementation, could you give an example on how to use your code?

**Comment by Mause** [comment]
Perhaps you could explain why what you want isn't a singleton?

```python

@singleton
def get_settings():
    return Settings()  # we only want one instance of this to exist


@app.get('/')
def index(settings = Depends(get_settings)):
     ....

````

**Comment by codethief** [comment]
Ok, looks like we were talking about different things here. A singleton to me is a class of which only one instance *can* exist. Of course there are many classes of which there only is one single instance at a given time but these are not necessarily singletons. Then again, this is only a matter of connotation.

In any case, thank you for your example! However, it doesn't really help me with my issue because I want `get_settings()` to return a different `Settings` object, depending on the chosen runtime profile etc. and for that I need proper dependency injection.

(Side remark: Why didn't you simply use a [@cache](https://docs.python.org/3/library/functools.html#functools.cache) decorator to avoid recreating the `Settings` instance?)

**Comment by Mause** [comment]
Yep, different meanings for different implementations (this is similar to spring boot's meaning of singleton).

Maybe you could expand what you mean by runtime profile?

I didn't use `@cache` as that would cache it globally within the python interpreter - this implementation essentially caches it per `FastAPI` instance, so if you create a new `FastAPI` instance per test, your tests won't leak state (for example).

**Comment by codethief** [comment]
> I didn't use @cache as that would cache it globally within the python interpreter - this implementation essentially caches it per FastAPI instance, so if you create a new FastAPI instance per test, your tests won't leak state (for example).

Oh, I see. That makes sense! :)

> Maybe you could expand what you mean by runtime profile?

In my particular case, the FastAPI process gets started by a parent process which starts the whole application stack. This parent process loads a configuration file, among others, and converts it into an instance of a certain `Configuration` class, which it then passes on to the web server via the `args` argument in `multiprocessing.Process()` like so:

```python
from dataclasses import dataclass
from multiprocessing import Process

from fastapi import FastAPI
import uvicorn

@dataclass
class Configuration:
    host: str
    port: int
    # …

# This function is invoked by the parent process
def start_webserver_process(config: Configuration):
    process = Process(
        target=webserver_process,
        name="Webserver process",
        args=(config, ),
    )
    process.start()

def webserver_process(config: Configuration):
    app = FastAPI()
    app.include_router(...)

    # Inject config object since some endpoints need it
    app.state.config = config

    uvicorn.run(
        app, host=config.host, port=config.port,
    )
```

As you can see, I cannot define a `get_settings()` singleton at "compilation" time as the configuration won't be known until runtime. I also don't want to set the config as module-global state in my web server process, as the state would leak during automated tests.

**Comment by Mause** [comment]
You could follow a similar pattern as my singleton function, like so:

```python


def get_configuration() -> Configuration:
    raise NotImplementedError("Configuration not injected")


def webserver_process(config: Configuration):
    ...

    app.dependency_overrides[get_configuration] = lambda: config

    ....
```

This lets you keep type safety as well

**Comment by codethief** [comment]
Yeah, I ended up doing something very similar:

```python
from typing import Type, TypeVar

from fastapi import FastAPI


T = TypeVar("T")


def inject_dependency(app: FastAPI, type_: Type[T], value: T):
    if not hasattr(app.state, "dependencies"):
        app.state.dependencies = {}

    if not isinstance(value, type_):
        raise ValueError("value needs to be an instance of type_")

    app.state.dependencies[type_] = value


def get_dependency(app: FastAPI, type_: Type[T]) -> T:
    value = app.state.dependencies[type_]
    if not isinstance(value, type_):
        raise ValueError(
            "The dependency that was found under the key type_ is not an instance of "
            "type_"
        )
    return value
```

Usage:

```python
from fastapi import FastAPI, Request

app = FastAPI()
config = Configuration()
inject_dependency(app, Configuration, config)

@app.get("/")
def get_device_state_endpoint(http_request: Request):
    config = get_dependency(http_request.app, Configuration)
    # …
```

Still, I think this should be an official feature.

**Comment by Mause** [comment]
With the above snippet, you can just inject it as a normal dependency (and you don't really need the function either)

```python

def webserver_process(config: Configuration):
    ...

    app.dependency_overrides[Configuration] = lambda: config

    ....


@app.get('/')
def index(configuration: Configuration = Depends()):
    ...

```

This is a working example:
```python

from fastapi.testclient import TestClient
from fastapi import FastAPI, Depends
from pydantic import BaseModel


class Config(BaseModel):
    name: str


app = FastAPI()
app.dependency_overrides[Config] = lambda: Config(name='hello')


@app.get('/')
def index(config: Config = Depends()):
    return config


tc = TestClient(app)
print(tc.get('/').json())

```

**Comment by gcasar** [comment]
@Mause this approach would inject the class as the API body  - Thats the main difference between your answer and what I understand @codethief was trying to achieve.

Is more similar to what FastAPI does for `Request` and the db session from the examples. Its not part of the API spec, its used to get the data / do business logic. As it stands the two are intertwined and you need to have a class that is not serializable.

**Comment by jonathanslenders** [comment, marked as answer]
I worked around this limitation by using Python's context variables.

A few lines of middleware will put them in place. The API endpoints can then either use some helper functions to get data from the context variables, possibly through ``fastapi.Depends``.

It's possible for the entry point of the application, e.g., the ``main()`` to wrap the ``ASGI`` app in middleware like this:
```python
from typing import Optional
from contextvars import ContextVar
from starlette.types import ASGIApp, Receive, Scope, Send

class ServerState:
    " Data to be exposed. Whatever would be global state otherwise. "

_server_state_var: ContextVar[Optional[ServerState]] = ContextVar("_server_state_var")

class ContextvarMiddleware:
    """
    Middleware that exposes the `ServerState` through contextvars to all
    underlying API endpoints.
    """
    def __init__(self, app: ASGIApp, server_state: ServerState) -> None:
        self.app = app
        self.server_state = server_state

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        _server_state_var.set(self.server_state)

        try:
            await self.app(scope, receive, send)
        finally:
            _server_state_var.set(None)

def create_api_handler(server_state: ServerState) -> ASGIApp:
    # To be called from "main()" or wherever the state is set.
    return ContextvarMiddleware(app, server_state=server_state)
```

With this, it's easy to define the ``FastAPI`` instance at the top level of a module and have it decoupled from any mutable state:

```python
from fastapi import FastAPI, Depends

def get_server_state() -> ServerState:
    server_state = _server_state_var.get()
    if server_state is None:
        raise RuntimeError("Application not wrapped in ContextvarMiddleware.")
    return server_state

api = FastAPI()

@api.get('/')
def home(server_state: ServerState = Depends(get_server_state)):
   ...
```

Personally, I prefer to run ASGI applications using Hypercorn for this, because it allows running ASGI applications programatically from a ``main()`` function, which is what you need if you have an installable entry point (with ``setup.py`` and maybe ``click`` for command line parsing).

@codethief: Would the above work for you?
@tiangolo: What do you think about this? Would it be possible or make sense to add this to the docs somewhere? (I don't think it's too much boilerplate to copy/paste for the people that need this functionality. Those who need it probably want to change it anyway, and will appreciate that it can be completely typed (which is maybe not the case if we try to come up with a generic solution.)

**Comment by codethief** [comment]
Hi @jonathanslenders , thanks so much for this interesting proposal!

To be honest, this is the first time I'm hearing of Python's contextvars module. Having now read the corresponding [PEP 567](https://www.python.org/dev/peps/pep-0567/), though, I'm getting the impression it's mainly meant for context-dependent state, i.e. state that's supposed to vary from execution context to execution context. In my case, however, I want to do the exact opposite: Namely, I'd like to inject the *same* global object / the same singleton into every execution context without having to actually resort to global state or the singleton pattern because it would obscure my application's true dependencies. While this would certainly work with contextvars, it requires the var to be set for every execution context separately which seems like a bit of overkill and, in case this is of any relevance, not very elegant.

PS: I believe in 

```python
    return ContextvarMiddleware(app, server_state=server_state)
```

you meant to write `self.app`?

**Comment by poolpitako** [comment]
is there any conclusion to this ticket?
What's the pattern recommended for injections outside of the request/app scope?

**Reply by AdoniasAlcantara** [reply]
It seems this discussion is dead :(
